### ISSUE TITLE
Adding DataView to WGSL to read various types of 1-byte-aligned values from a storage buffer.

### ISSUE THREAD

**Issue by m-schuetz** [gpuweb/gpuweb#2429]
# DataView for WGSL

I'd like to ask for/propose extending WGSL with an untyped storage buffer type via a DataView object that let's shaders read various types of attributes from arbitrary 1-byte-aligned locations of the storage buffer. WGSL has seen quite a few quality-of-life improvements over the past year, and I think this would also be a nice one. I'm frequently using custom readU8, readU16, readU32, etc. functions to achieve this, but having it as part of the standard would make things quite a bit easier.

### Why 

* Alignment rules of structs or even primitive types don't always match how 3D models or other data is stored in buffers. (e.g. not 4-byte-aligned)
* If stored in struct-of-arrays fashion, structs won't work at all and generating lots of individual bindings for each attribute complicates things.
* Data sets with dynamic sets of attributes and alignments may require developers to generate WGSL source code at runtime and handle bindings between JS and WGSL, further complicating things.
* It is often easier to simply bind an untyped buffer and let the shader do most of the interpreting of the data. 

### Example

Point clouds stored in [LAS record format 5 have a stride of 63 bytes](v), but [WebGPU requires that strides are a multiple of 4](https://www.w3.org/TR/webgpu/#vertex-formats). On top of that, all LAS record formats store RGB as "uint16x3", while WebGPU only accepts "uint16x2" and "uint16x4". So LAS data would either need to be transformed into a vertex-buffer friendly format at runtime, or we could use a WGSL-builtin DataView to load the data like this:

```
[[binding(0), group(0)]] var<storage, read> view : DataView;

fn main_vertex(vertex : VertexInput) -> VertexOutput {

	let stride = 63u;
	let x = view.getFloat32(stride * vertex.index + 0u);
	let y = view.getFloat32(stride * vertex.index + 4u);
	let z = view.getFloat32(stride * vertex.index + 8u);
	let R = view.getUint16(stride * vertex.index + 28u);
	let G = view.getUint16(stride * vertex.index + 30u);
	let B = view.getUint16(stride * vertex.index + 32u);

	...
}
```

Currently, I'm working around this issue with custom functions that read from a storage variable named buffer, such as:

```
fn readU8(offset : u32) -> u32{
	var ipos : u32 = offset / 4u;
	var val_u32 : u32 = buffer.values[ipos];
	var shift : u32 = 8u * (offset % 4u);
	var val_u8 : u32 = (val_u32 >> shift) & 0xFFu;

	return val_u8;
}

fn readU16(offset : u32) -> u32{
	var first = readU8(offset + 0u);
	var second = readU8(offset + 1u);
	var value = first | (second << 8u);

	return value;
}
```

### Potential issues

* Some types would need to be "up-casted", e.g. u8 and u16 to u32. 
* As far as I know, layout rules are sometimes there due to performance concerns? The workaround above works pretty great at least on desktop GPUs, though.

**Comment by kvark** [comment]
If we do this, would we do it any better than you are doing it already? I imagine on DX12 we can use a `ByteAddressBuffer`, but I don't recall anything similar in Vulkan or Metal. And even on DX12, since we always use `ByteAddressBuffer` internally, the compiler may as well end up producing exactly the same code, regardless of whether we are doing the packing/unpacking or the user does. With this in mind, it sounds like you can do all those things in user space today.

**Comment by m-schuetz** [comment]
Yeah, it would just be a quality of life improvement. You don't really need DataView in js either, but it makes things a whole lot more comfortable.

**Comment by magcius** [comment]
Note that ByteAddressBuffer addresses must be 4-byte aligned, it does not provide unaligned access. These restrictions are there for performance reasons, so I would like to encourage developers to stay on the fast path by default, and thus not provide these helpers. The user can still write utility methods that allow unaligned access.

**Comment by lifaon74** [comment]
As senior dev in js but a beginner in wgsl, I would love to have such feature, as I usually want to transfer u8 arrays of data instead of u32. Or have u8, u16 types direcly in wgsl. I know there are design and technical constraints behind u32 and f32 choices, but as an user perspective, I would love having such types.
